docs: reorganize content of readme.md (#1243)
microsoft/vscode-react-native: README.md

> [![Build status](https://dev.azure.com/vscode-webdiag-extensions/VS%20Code%20WebDiag%20extensions/_apis/build/status/vscode-react-native%20%5Bmaster%5D)](https://dev.azure.com/vscode-webdiag-extensions/VS%20Code%20WebDiag%20extensions/_build/latest?definitionId=2)

Reviewer: ```suggestion
[![Build status](https://dev.azure.com/vscode-webdiag-extensions/VS%20Code%20WebDiag%20extensions/_apis/build/status/vscode-react-native%20%5Bmaster%5D)](https://dev.azure.com/vscode-webdiag-extensions/VS%20Code%20WebDiag%20extensions/_build/latest?definitionId=60)
```

  - [Change project root](#change-project-root)
- [Developing inside a Docker Container](#developing-inside-a-docker-container)
- [Contributing](#contributing)
- [Known Issues](#known-issues)

# Getting started

Before going any further make sure that you:

* [have a working React Native environment](https://reactnative.dev/docs/environment-setup).
* are using [VS Code](https://code.visualstudio.com) and have [installed this extension from the Marketplace](https://marketplace.visualstudio.com/items?itemName=msjsdiag.vscode-react-native).
* have your React Native project root folder open in VS Code.

Please notice that the extension uses `.vscode/.react` directory at the project root to store intermediate files required for debugging. Although these files usually get removed after debug session ends, you may want to add this directory to your project's `.gitignore` file.

# React Native commands in the Command Palette

In the Command Palette, type `React Native` and choose a command.

![React Native commands](images/command-palette.png)

The **Run Android** command triggers `react-native run-android` and starts your app for Android.

> The **Run iOS** command similarly triggers `react-native run-ios` and starts your app in the iOS simulator (iPhone 6).

Reviewer 2: ```suggestion
The **Run iOS** command similarly triggers `react-native run-ios` and starts your app in the iOS simulator (e.g. iPhone 6).
```

The **Packager** commands allow you to start/stop the [**Metro Bundler**](https://github.com/facebook/metro-bundler) (formerly React Packager).

The full list of commands is:

> |Name|Description|

Reviewer: With that representation that table is quite poor. Command names just duplicate the descriptions. I think it could be useful if we expand descriptions a bit.

Author: I agree, but maybe not a blocker for this PR?
I'll open a separate issue to track this.

Author: I've opened #1246

Reviewer: Of course, I just mentioned this to make sure that we will do it in the future.

[![Build status](https://dev.azure.com/vscode-webdiag-extensions/VS%20Code%20WebDiag%20extensions/_apis/build/status/vscode-react-native%20%5Bmaster%5D)](https://dev.azure.com/vscode-webdiag-extensions/VS%20Code%20WebDiag%20extensions/_build/latest?definitionId=2)

This extension provides a development environment for React Native projects.
Using this extension, you can **debug your code and quickly run `react-native` commands** from the command palette.

![React Native features](images/react-features.gif)

<!-- TABLE OF CONTENTS -->

# Table of Contents

- [Getting started](#getting-started)
- [React Native commands in the Command Palette](#react-native-commands-in-the-command-palette)
- [Debugging React Native applications](#debugging-react-native-applications)
  - [Hermes](#hermes)
  - [iOS devices](#ios-devices)

>   - [Custmom scheme for iOS apps](#custom-scheme-for-ios-apps)

Author: It does not have to be a sub-heading of iOS. One is for real devices, the other is for iOS in general.

  - [TypeScript and Haul](#typescript-and-haul)
  - [Expo applications](#expo-applications)
    - [Configuring Expo](#configuring-expo)

>   - [React Native for Windows](#react-native-for-windows)

Author: Yeah, I had some mixed feeling about this. On one side we have iOS and Android support out of the box with some particular things for each one, then there's Expo, and finally we have Windows which is a bit more complicated to set up.
I put it also towards the end based on the current popularity. I'll give it a spin see what I come up with.

>   - [Debug configuration properties](#debug-configuration-properties)

Author: Debugger 👍

## Hermes

Hermes is the new JavaScript engine optimized for running React Native apps on Android. It improves app performance and decreases app size.

Click [here](https://reactnative.dev/docs/hermes) to learn more about Hermes.

Debugging apps with Hermes enabled is currently experimental. Please, see [this issue](https://github.com/microsoft/vscode-react-native/issues/1073) for current known issues on Hermes support.

To debug while using Hermes engine, please choose one of the following debug configurations:
 - React Native (Hermes): Debug Android - Experimental

```json
{
    "name": "Debug Android (Hermes) - Experimental",
    "cwd": "${workspaceFolder}",
    "type": "reactnativedirect",
    "request": "launch",
    "platform": "android"
}
```

 - React Native (Hermes): Attach to Hermes application - Experimental

```json
{
    "name": "Attach to Hermes application - Experimental",
    "cwd": "${workspaceFolder}",
    "type": "reactnativedirect",
    "request": "attach"
}
```

## iOS devices

Debugging on an iOS device require following manual steps:
* Install [ios-deploy](https://www.npmjs.com/package/ios-deploy) `npm install -g ios-deploy`.

> * Have a valid iOS Development certificate installed.

Reviewer 3: Should Development be capitalized here?

Author: ```suggestion
* Have a valid iOS development certificate installed.
```

This step was before in another example. We should be consistent with the wording. I believe it was:

> Install a valid iOS development certificate.

I'm OK either way. What do people prefer?

Reviewer 2: On the [official Apple page](https://developer.apple.com/support/certificates/), they often use the "Developer" with a capital letter, e.g. "Developer ID". So I suppose that the word "Development" could be capitalized.

* In your project's `launch.json` file set `target` to `device`. If you need to specify the exact device to run you can set `target` to `device=<iOS_device_name>` or you can also use `runArguments` property to specify particular device to run on in case of multiple devices connected, e.g. `"runArguments": [ "--device", "My iPhone" ]`

> * Choose **Debug iOS** configuration from the Configuration dropdown and press F5.

Reviewer 3: should Configuration be capitalized?

Author: ```suggestion
* Choose the **Debug iOS** option from the "Configuration" dropdown and press F5.
```

@jabfrei what do you think of this?

* Shake the device to open the development menu and select "Debug JS Remotely".

## Customs scheme for iOS apps

If you want to use custom scheme for your application you can either pass it as part of `runArguments` parameter arguments or set `scheme` configuration parameter as shown below:
```js
"runArguments": ["--scheme", "customScheme", ...]
// or
"runArguments": ["--scheme=customScheme", ...]
// or
"scheme" : "customScheme"
```
Please be aware, specifying scheme value as a part of `runArguments` parameter arguments will override `scheme` configuration parameter value if it set.

## TypeScript and Haul

> If you use Haul instead of the react-native packager, you must add `sourceMapPathOverrides` to the `launch.json` file.

Author: I'm thinking about linking to their website. What do you think?
Also we should probably mention the react-native packager is [Metro](https://facebook.github.io/metro/) 🤔

## Expo applications

To debug a project created using Expo or the create-react-native-app task, you can use embedded support for Expo.

Your environment must meet the following prerequisites:

- Install the [Expo app](https://getexponent.com/) on the target device or emulator
- Ensure that the react-native-cli is installed globally (`npm install -g react-native-cli`)

To start debugging in Expo follow these steps:

1. Open your project in VS Code with this extension installed.
1. Create a debug configuration (as described in [Debugging React Native applications](#debugging-react-native-applications)), select `Debug in Exponent` in the debug drop-down menu, and start debugging
1. Wait while some dependencies are configured - the extension will install `xde` and `@expo/ngrok` when this feature is used for the first time.
1. If you have not used exponent on this system before, you will be prompted for an exponent username and password.
   If you have not created an exponent account, then specifying a new username and password will create one.
   Note that there is no e-mail associated with the account, and no way to recover a forgotten password.

> 1. Once the packager starts, the extension will open a separate tab with QR code to scan from the Exponent app. Once you do so, the Exponent app will connect to the packager and begin running your app.

Reviewer 3: Expo, exponent, and Exponent are mixed here, should be consistent with case

> 1. Once the packager starts, the extension will open a separate tab with QR code to scan from the Exponent app. Once you do so, the Exponent app will connect to the packager and begin running your app.

Author: ```suggestion
1. If you have not used Exponent on this system before, you will be prompted for an Exponent username and password.
   If you have not created an Exponent account, then specifying a new username and password will create one.
   Note that there is no e-mail associated with the account, and no way to recover a forgotten password.
1. Once the packager starts, the extension will open a separate tab with QR code to scan from the Exponent app. Once you do so, the Exponent app will connect to the packager and begin running your app.
```

   ![React Native developer menu](./images/enable-remote-debug.png)

   From here you can run and debug the app as normal.

### Configuring Expo

The extension supports running through Exponent not just the applications created with Expo but even pure React Native applications (in that case you need to add `expo` package to `node_modules` in order to make it work with Expo: `npm install expo --save-dev`. In either cases it uses `app.json` configuration file in the root of the project.

If you are running `Debug in Exponent` configuration or any of pallette commands like `Run in Exponent`, `Publish to Exponent` then this file will be created automatically if absent or updated with the following basic configuration section:

```json
{
  "expo": {
    "slug": "MyApp", // Project slug
    "name": "MyApp", // Project name
    "sdkVersion": "31.0.0", // Expo SDK version
    "entryPoint": ".vscode\\exponentIndex.js" // Entrypoint for the project
  },
  "name": "MyApp" // Project name
}
```

> Full list of configurational parameters for `expo` section in `app.json` may be found on [official Expo documentation page](https://docs.expo.io/versions/latest/workflow/configuration).

Reviewer 3: should this be configuration over configurational?

Author: > should this be configuration over configurational?

configuration sounds better to me but English is not my first language so 🤷‍♂️

```suggestion
Full list of configuration parameters for `expo` section in `app.json` may be found on [official Expo documentation page](https://docs.expo.io/versions/latest/workflow/configuration).
```

> For running **pure React Native app**, extension, also, creates and uses `.vscode/exponentIndex.js` which is point to the app entrypoint (`index.js` or `index.android.js` or `index.ios.js`) file.

Reviewer 3: Grammar confusing here, not sure of all the terms here but this is best guess: "Running the **pure React Native app** extension also creates and uses `.vscode/exponentIndex.js`, which points to the app entrypoint (`index.js` or `index.android.js` or `index.ios.js`) file."

Author: ```suggestion
For running **pure React Native app**, the extension, creates and uses `.vscode/exponentIndex.js` which points to the app entrypoint (`index.js` or `index.android.js` or `index.ios.js`) file.
```

@SounD120 @RedMickey can you please double check this makes sense?

* Install a valid iOS development certificate.
* In your project's `launch.json` file set `target` to `device`. If you need to specify the exact device to run, you can set `target` to `device=<iOS_device_name>`, or you can also use `runArguments` property to specify a particular device to run on in case multiple devices are connected (e.g. `"runArguments": [ "--device", "My iPhone" ]`)
* Choose the **Debug iOS** option from the "Configuration" dropdown and press F5.
* Shake the device to open the development menu and select "Debug JS Remotely".

## Customs scheme for iOS apps

If you want to use a custom scheme for your application you can either pass it as part of the `runArguments` parameter arguments, or set the `scheme` configuration parameter as shown below:
```js
"runArguments": ["--scheme", "customScheme", ...]
// or
"runArguments": ["--scheme=customScheme", ...]
// or
"scheme" : "customScheme"
```
Please be aware, specifying the scheme value as a part of the `runArguments` parameter arguments will override the `scheme` configuration parameter value, if it set.

## Expo applications

> To debug a project created using Expo or the create-react-native-app task, you can use embedded support for Expo.

Reviewer 2: ```suggestion
To debug a project created using Expo or the `create-react-native-app` task, you can use embedded support for Expo.
```

Your environment must meet the following prerequisites:

- Install the [Expo app](https://getexponent.com/) on the target device or emulator

> - Ensure that the react-native-cli is installed globally (`npm install -g react-native-cli`)

Reviewer 2: ```suggestion
- Ensure that the `react-native-cli` is installed globally (`npm install -g react-native-cli`)
```

> - Ensure that the react-native-cli is installed globally (`npm install -g react-native-cli`)

Reviewer: ```suggestion
- Ensure that the react-native-cli is installed globally (`npm install -g @react-native-community/cli`)
```

Author: I believe this whole section is outdated. Installing globally the RN CLI is not recommended anymore and the steps to get [Expo up and running nowadays seem to be different](https://reactnative.dev/docs/environment-setup):

> Assuming that you have Node 12 LTS or greater installed, you can use npm to install the Expo CLI command line utility:
> `npm install -g expo-cli`
> Then run the following commands to create a new React Native project called "AwesomeProject":
> ```
> expo init AwesomeProject`
> cd AwesomeProject
> npm start # you can also use: expo start
> ```
> This will start a development server for you.

@SounD120 @RedMickey can you please confirm and tell me if there's any other requirements than the ones in the docs?

Reviewer: @molant Actually we might even don't need `expo-cli` at all for running Expo debugging scenarios since we install [`@expo/xdl`](https://www.npmjs.com/package/@expo/xdl) package inside the extension which is doing all the communication with Expo services. But it might be useful to init fresh Expo apps and to work with some of the extension commands. With respect to the React Native CLI - since [v0.13.0](https://github.com/microsoft/vscode-react-native/blob/master/CHANGELOG.md#0130) we can work with the locally installed CLI so the global one is not required.
@RedMickey correct me if I missed something.

Reviewer 2: Yep, It's possible to debug an Expo app without having `expo-cli` installed. I tried to do this and was able to debug an Expo app without problems.
Also now we don't use global react-native CLI.

Author: @RedMickey @SounD120 do you mind suggesting new instructions for this section?

Thanks!

Reviewer: Yes, it is better to update them

Author: @RedMickey @SounD120 what would be the new steps for this? I believe it's the last comment to get this merged 😊

Reviewer: I think we'll finally merge this and do all enhancements in another PRs. It was here for too long because of problems with CI tests, but now they gone, and I think we are good to go :)

Author: SGTM!
Will you take care of merging? Something else you need me to do?

Reviewer: I am waiting for the tests now, once they are passed I'll merge this.

To start debugging in Expo follow these steps:

1. Open your project in VS Code with this extension installed.
1. Create a debug configuration (as described in [Debugging React Native applications](#debugging-react-native-applications)), select `Debug in Exponent` in the debug drop-down menu, and start debugging
1. Wait while some dependencies are configured - the extension will install `xde` and `@expo/ngrok` when this feature is used for the first time.
1. If you have not used Exponent on this system before, you will be prompted for an Exponent username and password.

>    If you have not created an Exponent account, then specifying a new username and password will create one.

Reviewer 2: ```suggestion
   Exponent account allows you to use Expo cloud services. More info about how it works is available [here](https://docs.expo.io/versions/latest/workflow/how-expo-works/).
   If you have not created an Exponent account, then specifying a new username and password will create one.
```

>    Note that there is no e-mail associated with the account, and no way to recover a forgotten password.

Reviewer 2: ```suggestion
   Note that there is no e-mail associated with the account, and no way to recover a forgotten password.
   If you don't want to create an Exponent account, you can specify `expoHostType` parameter in your debug configuration to make Expo work locally (via LAN or on localhost).
```

|`scheme`|A scheme name to be passed to `react-native run-ios`, e.g. `devDebug` to specify `--scheme=devDebug`|`string`|n/a|

> |`productName`|iOS bundle display name e.g. `AwesomeProject` value means that extension will search for `AwesomeProject.app` bundle|`string`|n/a|

Reviewer 3: ...means that the extension will...

# Customization

> The extension can be further customized for other React Native scenarios, these are the most common ones:

Reviewer 3: ...scenarios. These are the most common:

            "group": "build",
            "type": "shell",
            "windows": {
                "command": "cd android; if($?) {./gradlew assembleDebug}"
            },
            "linux": {
                "command": "cd android && ./gradlew assembleDebug"
            }
        },
        {
            "taskName": "Build APK Release",
            "group": "build",
            "type": "shell",
            "windows": {
                "command": "cd android; if($?) {./gradlew assembleRelease}"
            },
            "linux": {
                "command": "cd android && ./gradlew assembleRelease"
            }
        },
        {
            "taskName": "Generate Android Bundle",
            "group": "build",
            "type": "shell",
            "command": "react-native bundle --platform android --dev false --entry-file index.js --bundle-output android/main.jsbundle"
        },
        {
            "taskName": "Generate iOS Bundle",
            "group": "build",
            "type": "shell",
            "command": "react-native bundle --platform ios --dev false --entry-file index.js --bundle-output ios/main.jsbundle"
        }
    ]
}
```

> To lear more about `tasks` in VSCode read [the official documentation](https://code.visualstudio.com/docs/editor/tasks).

Reviewer 3: To learn more...

> Visit [generating Signed APK](https://reactnative.dev/docs/signed-apk-android.html) to know more about this subject.

Reviewer 3: ... to learn more about...

## Specifying custom arguments for `react-native run-*` command

> For using custom run arguments for `react-native run-<platform>`:

Reviewer 3: Using custom run arguements...

**NOTE:** This overrides all other configuration parameters.

```json
{
  "react-native.android.runArguments.simulator": ["--appFolder", "/Users/test/AwesomeProject/android/app", "--deviceId", "emulator-5555"],
  "react-native.ios.runArguments.device": ["--project-path", "ios", "--device", "Max's iPhone"],
}
```

**NOTE:** You can get the list of installed simulator devices by:

> iOS devices(MacOS only):

Reviewer 3: iOS devices (macOS only):

```
xcrun simctl list --json devices
```

Android devices:

```
adb devices
```

> **NOTE:** If you want run application on iOS devices make sure you have `ios-deploy` installed globally.

Reviewer 3: ...want to run an application on iOS devices, make sure you...

```npm install -g ios-deploy```

## Setting up the react-native packager

To use a custom port for the `react-native` packager:

```json
{
  "react-native": {
    "packager" : {
      "port": portNumber
    }
  }
}
```

> If you change this port, then for iOS device and simulator scenarios you will have to modify the native code files. Instruction [here](https://blog.binoy.io/running-react-native-on-a-different-port-7deb43887cd4).

Reviewer 3: Instructions [here]

> If you use android, you need to change debug server by:

Reviewer 3: If you use Android, you need to change the debug server by:

1. CTRL+M(CMD+M) in the emulator
2. Go to Dev Settings
3. Debug server host for device => enter ‘localhost:\<yourPortNumber\>’.

> 4. Reload application (double R)

Reviewer 3: Is double R pressing R twice?

Author: Yes

**NOTE:** Some aspects of React Native hard-code the port to the default as specified in [this issue](https://github.com/facebook/react-native/issues/9145).

## Change project root

To specify a subfolder in which the react-native project is located, set `react-native-tools.projectRoot`. You can use either an absolute or relative path here:

```json
{
  "react-native-tools": {
    "projectRoot": "./your/react-native/project"
  }
}
```

# Developing inside a Docker Container

> The extension supports [VS Code Remote Development](https://code.visualstudio.com/docs/remote/remote-overview) features on Linux. Please follow the [VS Code official documentation](https://code.visualstudio.com/docs/remote/containers) to setup your environment to use remote development approach.

Reviewer 3: ...to use a remote development approach.

> You can use [official React Native Docker image](https://hub.docker.com/r/reactnativecommunity/react-native-android) provided by [react-native-community](https://github.com/react-native-community/docker-android).

Reviewer 3: ...provided by the [react-native-community]...

> Here are the steps to run React Native debugging inside Docker Container on a real Android device:

Reviewer 3: ...inside a Docker Container...

1. Open Command Palette and run the following command
    ```
    Remote-Containers: Add Development Container Configuration Files...
    ```
    Then select `Existing Dockerfile` to create `.devcontainer/devcontainer.json` configuration file.
1. Сreate Dockerfile extending [reactnativecommunity/react-native-android image](https://hub.docker.com/r/reactnativecommunity/react-native-android). For example you can use the following Dockerfile:
    ```
    FROM reactnativecommunity/react-native-android:latest

    RUN npm install -g expo-cli react-native-cli
    ```

> 1. Configure your `devcontainer.json` file just about like this: <br> **NOTE**: This is just a sample of configuration, you can modify your `devcontainer.json` file as you need.

Reviewer 3: 1. Configure your `devcontainer.json` file as needed. Below is an sample configuration:

        "context": "..",

        // Update the 'dockerFile' property if you aren't using the standard 'Dockerfile' filename.
        "dockerFile": "Dockerfile",

        // The optional 'runArgs' property can be used to specify additional runtime arguments.
        "runArgs": [
            "--privileged", // give all capabilities to a container, in other words, the container can then do almost everything that the host can do
            "--net", "host", // forwarding all host machine ports
            "-v", "/dev/bus/usb:/dev/bus/usb" // mount connected USB devices to a container
        ],

        "settings": {
            // This will ignore your local shell user setting for Linux since shells like zsh are typically
            // not in base container images. You can also update this to an specific shell to ensure VS Code
            // uses the right one for terminals and tasks. For example, /bin/bash (or /bin/ash for Alpine).
            "terminal.integrated.shell.linux": null
        },

        // Add the IDs of extensions you want installed when the container is created in the array below.
        "extensions": ["msjsdiag.vscode-react-native"]
    }
    ```

> 1. Open Command Palette and run the following command `Remote-Containers: Open Folder in Container` to reopen your project in container

Reviewer 3: ... your project in a container.

> 1. Connect your device via USB and start debugging the same way as on local machine

Reviewer 3: ...as on a local machine.

Currently the above scenario doesn't work on macOS and Windows. Docker Container implementation on these OS uses Virtual Machine tools which may have problems with USB forwarding for mobile devices.

# Contributing

> Please see our [contributing guide](CONTRIBUTING.md) for more information

Reviewer 3: ... for more information.

The extension supports [VS Code Remote Development](https://code.visualstudio.com/docs/remote/remote-overview) features on Linux. Please follow the [VS Code official documentation](https://code.visualstudio.com/docs/remote/containers) to setup your environment to use a remote development approach.

You can use [official React Native Docker image](https://hub.docker.com/r/reactnativecommunity/react-native-android) provided by the [react-native-community](https://github.com/react-native-community/docker-android).

Here are the steps to run React Native debugging inside a Docker Container on a real Android device:

1. Open Command Palette and run the following command
    ```
    Remote-Containers: Add Development Container Configuration Files...
    ```
    Then select `Existing Dockerfile` to create `.devcontainer/devcontainer.json` configuration file.
1. Сreate Dockerfile extending [reactnativecommunity/react-native-android image](https://hub.docker.com/r/reactnativecommunity/react-native-android). For example you can use the following Dockerfile:
    ```
    FROM reactnativecommunity/react-native-android:latest

    RUN npm install -g expo-cli react-native-cli
    ```

> 1. Configure your `devcontainer.json` file as needed. Below is an sample configuration:

Reviewer 2: ```suggestion
1. Configure your `devcontainer.json` file as needed. Below is a sample configuration:
```
